# DP: gm2 updates up to 20190817

git diff de2034bc023a89e45bec67e54d2a63982080c316 c67b2fd818227cd0d1a29051365b8e2a7a5efc02 | filterdiff --strip=2 --addoldprefix=a/src/ --addnewprefix=b/src/ --remove-timestamp

diffgcc/gm2/ChangeLog b/gcc-versionno/gcc/gm2/ChangeLog
index 43688ba0..49190553 100644
--- a/src/gcc/gm2/ChangeLog
+++ b/src/gcc/gm2/ChangeLog
@@ -1,4 +1,38 @@
-2019-08-05      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+2019-08-16      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/Make-lang.in:  (gm2/pge) added -lgcov.
+          (gm2/pg) added -lgcov.  (gm2/ppg)
+          added -lgcov.  (gm2/boot-bin/mc) added -lgcov.
+
+2019-08-13      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/Make-lang.in:  (check-gm2) removed, use inbuilt
+          version.  (lang_checks) include check-gm2.
+	* testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/examples/callingC/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/exceptions/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/extensions/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/imports/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/linking/libarchive/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/pim/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/types/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/lib/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+
+2019-08-13      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/lang-specs.h:  added %d to the temporary
+          assembly file.
+
+2019-08-12      Gaius Mulley <gaius.mulley@southwales.ac.uk>
 
 	* New branch gcc_9_2_0_gm2
 	* gm2/configure.in:  changed release number to 1.9.2.
diffgcc/gm2/Make-lang.in b/gcc-versionno/gcc/gm2/Make-lang.in
index 0af2ca63..85c006b1 100644
--- a/src/gcc/gm2/Make-lang.in
+++ b/src/gcc/gm2/Make-lang.in
@@ -1757,7 +1757,9 @@ gm2/mc-boot-ch/$(SRC_PREFIX)%.o: gm2/mc-boot-ch/$(SRC_PREFIX)%.c
 mc-bootstrap: $(GM2_DIRS) mc-clean gm2/boot-bin/mc$(exeext)
 
 gm2/boot-bin/mc$(exeext): $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) $(LDFLAGS) gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) $(LDFLAGS) \
+          gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o -lm \
+          -lgcov
 
 gm2/mc-boot/main.o: $(M2LINK) $(srcdir)/gm2/init/mcinit
 	unset CC ; $(M2LINK) -s --gcc --exit --name mainmcinit.c $(srcdir)/gm2/init/mcinit
@@ -2104,7 +2106,7 @@ gm2/gm2-ppg-boot/$(SRC_PREFIX)%.o: $(srcdir)/gm2/gm2-compiler/%.mod $(MCDEPS) $(
               -I$(srcdir)/gm2/mc-boot-ch -g -c gm2/gm2-ppg-boot/$(SRC_PREFIX)$*.c -o $@
 
 gm2/ppg$(exeext): gm2/boot-bin/mc $(BUILD-PPG-O) $(BUILD-MC-INTERFACE-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PPG-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PPG-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 
 gm2/gm2-ppg-boot/main.o: $(M2LINK) $(srcdir)/gm2/init/mcinit
 	unset CC ; $(M2LINK) -s --gcc --exit --name mainppginit.c $(srcdir)/gm2/init/ppginit
@@ -2150,7 +2152,7 @@ gm2/gm2-pg-boot/$(SRC_PREFIX)pg.o:  gm2/gm2-auto/pg.mod $(MCDEPS) $(BUILD-BOOT-H
 gm2/pg$(exeext): gm2/boot-bin/mc \
     $(BUILD-PG-O) $(GM2-PPG-MODS:%.mod=gm2/gm2-pg-boot/%.o) \
     $(BUILD-MC-INTERFACE-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PG-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PG-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 
 gm2/gm2-auto/pginit:
 	sed -e 's/ppg/pg/' < $(srcdir)/gm2/init/ppginit > $@
@@ -2213,7 +2215,7 @@ gm2/gm2-pge-boot/$(SRC_PREFIX)pge.o:  gm2/gm2-auto/pge.mod $(MCDEPS) $(BUILD-BOO
 gm2/pge$(exeext): gm2/boot-bin/mc \
     $(BUILD-PGE-O) $(GM2-PPG-MODS:%.mod=gm2/gm2-pge-boot/%.o) \
     $(BUILD-MC-INTERFACE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PGE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PGE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 	$(srcdir)/gm2/tools-src/buildpg $(srcdir)/gm2/gm2-compiler/ppg.mod t > gm2/gm2-auto/t.bnf
 	./gm2/pge$(exeext) gm2/gm2-auto/t.bnf > gm2/gm2-auto/t1.mod
 	./gm2/pg$(exeext) gm2/gm2-auto/t.bnf > gm2/gm2-auto/t2.mod
@@ -2414,17 +2416,7 @@ check-modula-2: check-gm2
 check_modula-2: check-gm2
 check_modula-2: check-gm2
 
-GM2TESTSUITEDIR=testsuite
-
-check-gm2: $(GM2TESTSUITEDIR)/site.exp
-	-(rootme=`${PWD_COMMAND}`; export rootme; \
-	srcdir=`cd ${srcdir}; ${PWD_COMMAND}` ; export srcdir ; \
-	cd $(TESTSUITEDIR); \
-	EXPECT=${EXPECT} ; export EXPECT ; \
-	if [ -f $${rootme}/../expect/expect ] ; then  \
-	   TCL_LIBRARY=`cd .. ; cd ${srcdir}/../tcl/library ; ${PWD_COMMAND}` ; \
-	    export TCL_LIBRARY ; fi ; \
-	$(RUNTEST) --tool gm2 $(RUNTESTFLAGS))
+lang_checks += check-gm2
 
 check-gm2-local: $(GM2TESTSUITEDIR)/site.exp
 	-(rootme=`${PWD_COMMAND}`; export rootme; \
diffgcc/gm2/lang-specs.h b/gcc-versionno/gcc/gm2/lang-specs.h
index ba961343..fc7dae2a 100644
--- a/src/gcc/gm2/lang-specs.h
+++ b/src/gcc/gm2/lang-specs.h
@@ -31,7 +31,7 @@ Boston, MA 02110-1301, USA.  */
                        %{MT*} %{MF*} -quiet "
 
 #define GM2CC(INPUT,OUTPUT) \
-  "%{!fno-exceptions:cc1plus;:cc1} " GM2CC_OPTIONS " " INPUT " -o %b_m2.s \n\
+  "%{!fno-exceptions:cc1plus;:cc1} " GM2CC_OPTIONS " " INPUT " -o %d%b_m2.s \n\
   " AS("%b_m2.s",OUTPUT) " "
 
 #define GM2LCC(OBJECT,LST) \
diffgcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp
index 0eb58402..923ceafd 100644
--- a/src/gcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp
@@ -27,12 +27,12 @@ set gm2src ${srcdir}/../gm2
 
 gm2_init_pim "${srcdir}/gm2/calling-c/datatypes/unbounded/run/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with "c.o"
 
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/calling-c/datatypes/unbounded/run/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/calling-c/datatypes/unbounded/run/pass/c.c -o c.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
diffgcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp
index 6d0a31f5..c1528ccd 100644
--- a/src/gcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -26,11 +26,11 @@ set gm2src ${srcdir}/../gm2
 
 gm2_init_iso "$srcdir/gm2/examples/callingC/run/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with c.o
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/examples/callingC/run/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/examples/callingC/run/pass/c.c -o c.o]
 
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
diffgcc/testsuite/gm2/exceptions/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/exceptions/run/pass/gm2.exp
index 6f43cc8c..ee0b11d3 100644
--- a/src/gcc/testsuite/gm2/exceptions/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/exceptions/run/pass/gm2.exp
@@ -25,11 +25,11 @@ load_lib prune.exp
 load_lib gm2-torture.exp
 
 
-set XGPP [lookfor_file $tmpdir xg++];
+set XGPP [lookfor_file $rootme xg++];
 
 set output [exec rm -f cpp.o mycpp.o]
-set output [exec ${XGPP} -B[file dirname $tmpdir] -g -c $srcdir/gm2/exceptions/run/pass/cpp.cpp]
-set output [exec ${XGPP} -B[file dirname $tmpdir] -g -c $srcdir/gm2/exceptions/run/pass/mycpp.cpp]
+set output [exec ${XGPP} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/exceptions/run/pass/cpp.cpp]
+set output [exec ${XGPP} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/exceptions/run/pass/mycpp.cpp]
 
 #
 #  notice this uses PIM libraries with exceptions - this is a useful test.
diffgcc/testsuite/gm2/extensions/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/extensions/run/pass/gm2.exp
index 492ef7d9..4b632030 100644
--- a/src/gcc/testsuite/gm2/extensions/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/extensions/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -28,10 +28,10 @@ gm2_init_pim "${srcdir}/gm2/extensions/run/pass" -fsoft-check-all
 
 gm2_link_with cvararg.o
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 set output [exec rm -f cvararg.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/extensions/run/pass/cvararg.c -o cvararg.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/extensions/run/pass/cvararg.c -o cvararg.o]
 
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
diffgcc/testsuite/gm2/imports/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/imports/run/pass/gm2.exp
index eea64f1c..1996b26d 100644
--- a/src/gcc/testsuite/gm2/imports/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/imports/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -29,7 +29,7 @@ gm2_init_pim "${srcdir}/gm2/imports/run/pass"
 gm2_link_with "-lgm2"
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
-    set output [exec ../xgm2 -B../ -g -c "-I${gccpath}/libgm2/libpim:${gm2src}/gm2-libs:${srcdir}/gm2/imports/run/pass" -fpim ${srcdir}/gm2/imports/run/pass/c.mod]
+    set output [exec ../../xgm2 -B../../ -g -c "-I${gccpath}/libgm2/libpim:${gm2src}/gm2-libs:${srcdir}/gm2/imports/run/pass" -fpim ${srcdir}/gm2/imports/run/pass/c.mod]
 
     # If we're only testing specific files and this isn't one of them, skip it.
     if ![runtest_file_p $runtests $testcase] then {
diffgcc/testsuite/gm2/linking/libarchive/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/linking/libarchive/pass/gm2.exp
index a7b84536..0f6217aa 100644
--- a/src/gcc/testsuite/gm2/linking/libarchive/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/linking/libarchive/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -24,13 +24,13 @@ load_lib gm2-torture.exp
 
 set gm2src ${srcdir}/../gm2
 
-gm2_init_iso "${srcdir}/gm2/linking/libarchive/pass" 
+gm2_init_iso "${srcdir}/gm2/linking/libarchive/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with c.o
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/linking/libarchive/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/linking/libarchive/pass/c.c -o c.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
diffgcc/testsuite/gm2/pim/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/pim/run/pass/gm2.exp
index 74cd1fec..3f571f90 100755
--- a/src/gcc/testsuite/gm2/pim/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/pim/run/pass/gm2.exp
@@ -27,7 +27,7 @@ set gm2src ${srcdir}/../gm2
 gm2_init_pim "${srcdir}/gm2/pim/run/pass" -fsoft-check-all
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
-    set output [exec ../xgm2 -B../ -g -c -I$srcdir/../gm2/gm2-libs -I$srcdir/gm2/pim/run/pass -I$srcdir/../gm2/gm2-compiler -I../gm2/gm2-libs -I../gm2/gm2-compiler -fpim $srcdir/gm2/pim/run/pass/sys.mod]
+    set output [exec ../../xgm2 -B../../ -g -c -I$srcdir/../gm2/gm2-libs -I$srcdir/gm2/pim/run/pass -I$srcdir/../gm2/gm2-compiler -I../gm2/gm2-libs -I../gm2/gm2-compiler -fpim $srcdir/gm2/pim/run/pass/sys.mod]
 
     # If we're only testing specific files and this isn't one of them, skip it.
     if ![runtest_file_p $runtests $testcase] then {
diffgcc/testsuite/gm2/types/run/pass/gm2.exp b/gcc-versionno/gcc/testsuite/gm2/types/run/pass/gm2.exp
index bef297d7..ddc8c7dd 100755
--- a/src/gcc/testsuite/gm2/types/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/types/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -24,14 +24,14 @@ load_lib gm2-torture.exp
 
 set gm2src ${srcdir}/../gm2
 
-gm2_init_pim "${srcdir}/gm2/types/run/pass" 
+gm2_init_pim "${srcdir}/gm2/types/run/pass"
 
 gm2_link_with d.o
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 set output [exec rm -f d.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/types/run/pass/d.c -o d.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/types/run/pass/d.c -o d.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
@@ -42,4 +42,4 @@ foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     gm2-torture-execute $testcase "" "pass"
 }
 
-set output [exec rm -f d.o]
\ No newline at end of file
+set output [exec rm -f d.o]
diffgcc/testsuite/lib/gm2.exp b/gcc-versionno/gcc/testsuite/lib/gm2.exp
index fd6b44e7..791473df 100755
--- a/src/gcc/testsuite/lib/gm2.exp
+++ b/src/gcc/testsuite/lib/gm2.exp
@@ -92,8 +92,8 @@ proc gm2_init { args } {
     if { $gm2_initialized == 1 } { return; }
 
     set gm2_link_libraries "";
-    set GCC_UNDER_TEST [lookfor_file $tmpdir xgm2];
-    append GCC_UNDER_TEST " " -B[file dirname $tmpdir] " " ${args};
+    set GCC_UNDER_TEST [lookfor_file $rootme xgm2];
+    append GCC_UNDER_TEST " " -B[file dirname $rootme]/gcc " " ${args};
     append GCC_UNDER_TEST " " -fno-diagnostics-show-caret
     append GCC_UNDER_TEST " " -fno-diagnostics-show-line-numbers
     append GCC_UNDER_TEST " " -fdiagnostics-color=never
@@ -123,7 +123,6 @@ proc gm2_init { args } {
 #
 
 proc gm2_target_compile_default { source dest type options } {
-    global tmpdir;
     global gluefile wrap_flags;
     global GCC_UNDER_TEST;
     global TOOL_OPTIONS;
